### PR TITLE
Vlan snmp

### PR DESCRIPTION
**Pull request by raphaelt-nvidia**
**- What I did**

Restored snmp vlan support per RFC1213, which had been reverted due to inconsistency with RFC2863, and added the missing support for RFC2863.  Added unit tests for RFC2863.

**- How I did it**

Reverted https://github.com/Azure/sonic-snmpagent/pull/191, which was itself a revert of https://github.com/Azure/sonic-snmpagent/pull/169.  Then added code to support vlan in rfc2863.py and unit tests.

Since a long time has elapsed since the original commit and other code has been pushed in the meantime, great care is needed in merging.  Note in particular that mibs.init_sync_d_lag_tables now returns five parameters, the last two of which were added: lag_sai_map and sai_lap_map.  This PR needs one of those maps, and a concurrent commit supporting RFC4363 needs the other, so all the callers were updated and use the one they need.

**- How to verify it**

Unit tests are run via make target/python-wheels/asyncsnmp-2.1.0-py3-none-any.whl.
See also https://github.com/Azure/sonic-snmpagent/pull/169.

**- Description for the changelog**

Restored support for aggregated router interface counters and L3 VLAN counters to RFC1213 MIB, and extended to RFC2863.

### CODE REVIEW

File: tests/test_mibs.py [hunk 1]
```diff
@@ -14,14 +14,19 @@ class TestGetNextPDU(TestCase):
...
         db_conn = Namespace.init_namespace_dbs()
 
         lag_name_if_name_map, \
         if_name_lag_name_map, \
+<<<<<<< HEAD
+        oid_lag_name_map, \
+        lag_sai_map = Namespace.get_sync_d_from_all_namespace(mibs.init_sync_d_lag_tables, db_conn)
+=======
         oid_lag_name_map, _ = Namespace.get_sync_d_from_all_namespace(mibs.init_sync_d_lag_tables, db_conn)
+>>>>>>> 4e063e4ade89943f2413a767f24564aecfa2cd1c
```
Comment (stepanblyschak) [inline tests/test_mibs.py]: Remove git conflict markers

File: tests/test_mibs.py [hunk 2]
```diff
@@ -30,3 +35,4 @@ def test_init_sync_d_lag_tables(self):
 
         self.assertTrue("PortChannel_Temp" in lag_name_if_name_map)
         self.assertTrue(lag_name_if_name_map["PortChannel_Temp"] == [])
+        self.assertTrue(lag_sai_map["PortChannel01"] == "2000000000006")
```
Comment (stepanblyschak) [inline tests/test_mibs.py:34]: What is "2000000000006"?

File: src/sonic_ax_impl/mibs/ietf/rfc2863.py [hunk 3]
```diff
@@ -103,6 +149,41 @@ def update_data(self):
...
+
+    def update_rif_counters(self):
+        rif_sai_ids = list(self.rif_port_map) + list(self.vlan_name_map)
+        for sai_id in rif_sai_ids:
+            counters_db_data = Namespace.dbs_get_all(self.db_conn, mibs.COUNTERS_DB,
+                                                     mibs.counter_table(mibs.split_sai_id_key(sai_id)[1]),
+                                                     blocking=False)
+            self.rif_counters[sai_id] = {
+                counter: int(value) for counter, value in counters_db_data.items()
+            }
```
Comment (stepanblyschak) [inline src/sonic_ax_impl/mibs/ietf/rfc2863.py]: This seems to be duplicated code from rfc1213. Can we write this logic once and reuse in both mibs?

File: src/sonic_ax_impl/mibs/ietf/rfc2863.py [hunk 4]
```diff
@@ -153,6 +236,31 @@ def interface_alias(self, sub_id):
 
         return entry.get("description", "")
 
+    def aggregate_counters(self):
```
Comment (stepanblyschak) [inline src/sonic_ax_impl/mibs/ietf/rfc2863.py]: Same comment regarding code duplication

File: tests/namespace/test_interfaces.py [hunk 5]
```diff
@@ -367,6 +367,495 @@ def test_in_octets_override(self):
...
+        )
+
+        response = get_pdu.make_response(self.lut)
+        print(response)
+
+        value0 = response.values[0]
+        self.assertEqual(value0.type_, ValueType.COUNTER_32)
+        self.assertEqual(str(value0.name), str(ObjectIdentifier(11, 0, 1, 0, (1, 3, 6, 1, 2, 1, 2, 2, 1, 20, 1001))))
+        self.assertEqual(value0.data, 106)
+
```
Comment (stepanblyschak) [inline tests/namespace/test_interfaces.py:862]: Can you add similar tests for rfc2863?

**Review by stepanblyschak — CHANGES_REQUESTED** [review]

File: tests/test_mibs.py (hunk 1 above)
Comment (raphaelt-nvidia) [reply tests/test_mibs.py]: Missed that.  Taking the HEAD branch as in https://github.com/Azure/sonic-snmpagent/pull/169/files.

File: tests/test_mibs.py (hunk 2 above)
Comment (raphaelt-nvidia) [reply tests/test_mibs.py:34]: It was added in https://github.com/Azure/sonic-snmpagent/pull/169.

File: src/sonic_ax_impl/mibs/ietf/rfc2863.py (hunk 3 above)
Comment (raphaelt-nvidia) [reply src/sonic_ax_impl/mibs/ietf/rfc2863.py]: I can try, but to do that I need to understand the code better in order to write it in an understandable way.  In update_data(), the following code was deleted in the original commit (PR 169) from rfc1213, but retained in rfc2863.  Do you know why, and what the code does?

File: tests/test_mibs.py (hunk 2 above)
Comment (stepanblyschak) [reply tests/test_mibs.py:34]: Resolved

File: src/sonic_ax_impl/mibs/ietf/rfc2863.py (hunk 3 above)
Comment (raphaelt-nvidia) [reply src/sonic_ax_impl/mibs/ietf/rfc2863.py]: Basim and Stepan agreed that counters are not needed in rfc2863, therefore I have deleted update_if_counters(), update_rif_counters(), aggregate_counters(), and the calls to them.

File: src/sonic_ax_impl/mibs/ietf/rfc2863.py (hunk 4 above)
Comment (raphaelt-nvidia) [reply src/sonic_ax_impl/mibs/ietf/rfc2863.py]: See above.

File: tests/namespace/test_interfaces.py (hunk 5 above)
Comment (raphaelt-nvidia) [reply tests/namespace/test_interfaces.py:862]: Again, after speaking with Basim, RFC2863 counters are not required, so I looked for vlan references in the file.  The only reference is in test_in_ucast_vlan_subinterface(), which is a counter, so I don't see what additional tests I should write.

**Comment by liat-grozovik** [comment]
@raphaelt-nvidia have you upstream this already? i could not find an open PR so it is either not upstream or merged. if merged, lets close this one, if not, lets please have it done.

**Pull request closed without merging**